fendermint/testing/materializer/src/docker/node.rs

```diff
@@ -0,0 +1,566 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    os::unix::fs::MetadataExt,
+    path::{Path, PathBuf},
+};
+
+use anyhow::{anyhow, Context};
+use bollard::{
+    container::{Config, CreateContainerOptions, RemoveContainerOptions},
+    service::{HostConfig, PortBinding},
+    Docker,
+};
+use ethers::types::H160;
+use lazy_static::lazy_static;
+
+use super::{
+    container::DockerContainer, DockerConstruct, DockerMaterials, DockerNetwork, DockerPortRange,
+    DockerWithDropHandle,
+};
+use crate::{
+    docker::DOCKER_ENTRY_FILE_NAME,
+    materializer::{NodeConfig, TargetConfig},
+    materials::export_file,
+    NodeName, ResourceHash,
+};
+
+// TODO: Add these to the materializer.
+const COMETBFT_IMAGE: &str = "cometbft/cometbft:v0.37.x";
+const FENDERMINT_IMAGE: &str = "fendermint:latest";
+
+const STATIC_ENV: &str = "static.env";
+const DYNAMIC_ENV: &str = "dynamic.env";
```

Comment:
I am slightly confused by the difference between `STATIC_ENV` and  `DYNAMIC_ENV`. Why do we need a dynamic.env file instead of just passing them at runtime, is it a docker thing? Also, didn't find any cases of env DYNAMIC_ENV variables, is that expected?

Reply:
Thanks for pointing this out, I'll leave a comment about it here. I mentioned it somewhere else but on their own they are confusing. The dynamic env only comes into play in https://github.com/consensus-shipyard/ipc/pull/744

So the idea was that we have `Materializer::create_node` and `Materializer::start_node` separately, and the only reason for this is to support the use case where we have node A bootstrapping from node B, and at the same time node B bootstrapping from A. This looks like a contradiction, but often the P2P layer allows things to kick off and wait for the other node to appear. This makes sense to me as I can operate 2 seeds: start the 1st, then the 2nd connecting to it, then if I need to restart the 1st, it connects to the 2nd. I think CometBFT persists previous connections, so maybe allowing this is an overkill. 

But there is another use case: in our current scripts, we create a fully connected network (every node to every other), and this requires knowing not just the address of other peers, but their peer IDs (the hash of their network keys). 

In my model the `create_node` is responsible for setting up everything, which includes the network key. If it started the node as well, then the nodes would have to form a DAG, only connecting to already created and started nodes, because that's when you can already look up the network ID. Instead, we first create everything, then start everything, and in that second pass, everyone's IDs are already known, so anyone can connect to anybody else as they please. 

And so, `static.env` is what can be put into file during creation, while `dynamic.env` is what is only apparent when the node starts. I could have just appended more values to a file, but thought it's easier if we restart nodes to just write it to a different file, which can thus be overwritten.